File: 1753.py
import sys
import heapq
INF = int(1e9)
input = sys.stdin.readline

v,e = map(int,input().split())
k = int(input())

graph = [[] for _ in range(v+1)]
distance = [INF]*(v+1)
distance[k] = 0
for _ in range(e):
    a,b,c = map(int,input().split())
    graph[a].append([b,c])

def dijkstra(k):
    q = []
    heapq.heappush(q,[0,k])
    while q:
        dist,now = heapq.heappop(q)
        if distance[now]<dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost<distance[i[0]]:
                distance[i[0]]=cost
                heapq.heappush(q,[cost,i[0]])

dijkstra(k)
for i in range(1,v+1):
    if distance[i]==INF:
        print("INF")
    else:
        print(distance[i])

File: BOJ/1753.js
let fs = require('fs');
let dev = true;
let input = fs
  .readFileSync(dev ? 'test.txt' : './dev/stdin')
  .toString()
  .trim()
  .split('\n');
let [v, e] = input.shift().split(' ').map(Number);
let s = Number(input.shift());
input = input.map((e) => e.split(' ').map(Number));
let board = Array.from({ length: v + 1 }, () => new Array());
input.forEach((e) => {
  let [start, end, w] = e;
  board[start].push([end, w]);
});

let dp = new Array(v + 1).fill(Infinity);
const dijstra = (s) => {
  let q = [[0, s]];
  dp[s] = 0;
  while (q.length) {
    let [distance, node] = q.shift();
    if (dp[node] < distance) continue;
    for (const next of board[node]) {
      let cost = distance + next[1];
      if (cost < dp[next[0]]) {
        dp[next[0]] = cost;
        q.push([cost, next[0]]);
        q.soft((a, b) => a[0] - b[0]);
      }
    }
  }
};

// const dp = new Array(v + 1).fill(Infinity);

// const dijkstra = (s) => {
//   let q = [];
//   q.push([0, s]);
//   dp[s] = 0;
//   while (q.length > 0) {
//     let [distance, cur] = q.shift();
//     if (dp[cur] < distance) continue;
//     for (const next of board[cur]) {
//       let cost = distance + next[1];
//       if (dp[next[0]] > cost) {
//         dp[next[0]] = cost;
//         q.push([cost, next[0]]);
//         q.sort((a, b) => a[0] - b[0]);
//       }
//     }
//   }
// };
// dijkstra(s);
// for (let i = 1; i < v + 1; i++) {
//   dp[i] == Infinity ? console.log('INF') : console.log(dp[i]);
// }
